### PR TITLE
[refactor][Relations]: state management (removing connect / disconnect)

### PR DESCRIPTION
**Pull request by joshuaellis**
### What does it do?

* `RelationsInputDataManager` now uses `modifiedData` as it's source of truth for the UI
* Diffs relations at the necessary points (cleanData / searching)
* Removes the concept of `connect` and `disconnect` arrays from the store

The state management model is now like this:
image

### Why is it needed?

* Simplifies state management in preparation for re-ordering
* Solves flashes of old content between server sync updates
* Comments the code which is good for the soul 👀
* Sets up resolving CONTENT-546

### How to test it?

Unit tests have been updated

### Related issue(s)/PR(s)

* CONTENT-619
* CONTENT-545
* CONTENT-623

### In collaboration with

* @ronronscelestes

### CODE REVIEW

**Comment by joshuaellis** [comment]
@gu-stav not ready yet, unit tests will fail, resolving those now but wanted to show you the work in advance 👍🏼

**Comment by gu-stav** [comment]
@joshuaellis Wow, that looks epic. Thanks for sharing. Let me know once you are ready for a first review!

**Comment by joshuaellis** [comment]
@ronronscelestes & @gu-stav you can now review the code. I'm working on the `todo` tests and fixing ones i've broken in the reducer but I don't expect the code i've written to change (from my POV).

I think its most important to me that it's understandable what's going on, so if something is unclear, if you think it would benefit from comments lmk. Same with tests, if i've forgotten an edge case then do let me know as well.

**Review by ronronscelestes — COMMENTED** [review]
Thank you for this huge work 💃

Tested a schema with relation in component and nested component, found an error there:
image

I have a component NestedCompRel which contains a component CompRel which contains a relation
image
I then used this component in a CT to create single component, rep component and dynamic-zone.
If I add a relation to the first field then save, then to the next one, save, etc - everything works well.
But if I try to set relations for each fields at once and then save, that's when the error appears!
Hope it helps 🙏

**Comment by joshuaellis** [comment]
> Thank you for this huge work 💃
> 
> Tested a schema with relation in component and nested component, found an error there: image
> 
> I have a component NestedCompRel which contains a component CompRel which contains a relation image I then used this component in a CT to create single component, rep component and dynamic-zone. If I add a relation to the first field then save, then to the next one, save, etc - everything works well. But if I try to set relations for each fields at once and then save, that's when the error appears! Hope it helps 🙏

I have tests for relations inside components inside components, so im not sure why this has failed.

File: packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/index.js [hunk 1]
```diff
@@ -170,12 +203,15 @@ const EditViewDataManagerProvider = ({
     });
   }, []);
 
-  const connectRelation = useCallback(({ target: { name, value, replace } }) => {
+  /**
+   * @type {({ name: string, value: Relation, isSingleRelation: boolean}) => void}
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/index.js:207]: The name of the param `isSingleRelation` is out of sync with the actual function.

File: packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/reducer.js [hunk 2]
```diff
@@ -34,212 +40,177 @@ const reducer = (state, action) =>
...
-           * The content API only returns { count: <int> }, which is why
-           * we need to extend the existing state rather than overwriting it.
-           */
+        /**
+         * You can't mutate an actions value.
+         * and spreading an object only clones
+         * the first level, the deeply nested values
+         * are a reference.
+         */
+        const data = cloneDeep(initialValues);
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/reducer.js:173]: FYI: In the near future all browsers will have [`structuredClone()`](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone).

File: packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/tests/reducer.test.js [hunk 3]
```diff
@@ -539,112 +727,82 @@ describe('CONTENT MANAGER | COMPONENTS | EditViewDataManagerProvider | reducer',
...
         initialData: {
-          relation: { results: [{ id: 2 }, { id: 1 }] },
+          relation: [{ id: 2 }, { id: 1 }],
         },
         modifiedData: {
-          relation: {
-            connect: [],
-            disconnect: [],
-          },
+          relation: [{ id: 2 }, { id: 1 }],
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/tests/reducer.test.js:733]: I'm wondering: does that mean for a user (plugin-developer) it will be impossible to differentiate which relations have changed, because `initialData` and `modifiedData` are always the same? That would be a breaking change imo and defeat the idea of `modifiedData`. At the moment this is the only integration point we give to plugin developers if they want to interfact with relational fields in the UI and therefore I think we shouldn't break it.

File: packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/utils/cleanData.js [hunk 4]
```diff
@@ -4,14 +4,31 @@ import isObject from 'lodash/isObject';
 
 /* eslint-disable indent */
 
-const cleanData = (retrievedData, currentSchema, componentsSchema) => {
+/**
+ *
+ * @param {{ browserState: object, serverState: object }} browserState – the modifiedData from REDUX, serverStaet - the initialData from REDUX
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/utils/cleanData.js:9]:
```suggestion
 * @param {{ browserState: object, serverState: object }} browserState – the modifiedData from REDUX, serverState - the initialData from REDUX
```

File: packages/core/admin/admin/src/content-manager/components/RelationInput/RelationInput.js [hunk 5]
```diff
@@ -347,36 +345,29 @@ const RelationInput = ({
 };
 
 const ReactQueryRelationResult = PropTypes.shape({
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/RelationInput/RelationInput.js:347]: I think we shouldn't call the type any longer `ReactQueryRelationResult`, because it doesn't contain a full react-query response. What about

```suggestion
const RelationsResult = PropTypes.shape({
```

The reason why I wanted to pass in a paginated result is, that loadMore should handle scrolling at some point (users clicks "load more" -> scroll to the end of the fetched page), but I guess we could start to pass the pagination information on a different way.

File: packages/core/admin/admin/src/content-manager/components/RelationInput/RelationInput.js [hunk 6]
```diff
@@ -347,36 +345,29 @@ const RelationInput = ({
...
+      mainField: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+    })
+  ),
   hasNextPage: PropTypes.bool,
   isFetchingNextPage: PropTypes.bool.isRequired,
   isLoading: PropTypes.bool.isRequired,
   isSuccess: PropTypes.bool.isRequired,
 });
 
 const ReactQuerySearchResult = PropTypes.shape({
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/RelationInput/RelationInput.js:362]: Same here. 
```suggestion
const SearchResults = PropTypes.shape({
```

File: packages/core/admin/admin/src/content-manager/components/RelationInputDataManager/RelationInputDataManager.js [hunk 7]
```diff
@@ -34,14 +38,26 @@ export const RelationInputDataManager = ({
...
+    name,
     relation: {
       enabled: get(initialData, name)?.count !== 0 && !!endpoints.relation,
       endpoint: endpoints.relation,
+      pageGoal: currentLastPage,
       pageParams: {
         ...defaultParams,
         pageSize: RELATIONS_TO_DISPLAY,
       },
+      onLoadRelationsCallback: loadRelation,
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/RelationInputDataManager/RelationInputDataManager.js:55]: I'd suggest to change the method name, as it is already scoped to relations.

```suggestion
      onLoad: loadRelation,
```

File: packages/core/admin/admin/src/content-manager/components/RelationInputDataManager/RelationInputDataManager.js [hunk 8]
```diff
@@ -106,30 +92,39 @@ export const RelationInputDataManager = ({
...
     });
   };
 
-  const handleOpenSearch = () => {
-    searchFor('', {
-      idsToInclude: relationsFromModifiedData?.disconnect?.map((relation) => relation.id),
-      idsToOmit: relationsFromModifiedData?.connect?.map((relation) => relation.id),
-    });
-  };
+  const handleOpenSearch = () => handleSearch();
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/RelationInputDataManager/RelationInputDataManager.js:127]: Could we simplify this then and not pass two props to `RelationInput`?

File: packages/core/admin/admin/src/content-manager/components/RelationInputDataManager/utils/diffRelations.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+ * @param {Array<{id: string}>} browserStateRelations
+ * @param {Array<{id: string}>} serverStateRelations
+ * @returns {[connected: string[], disconnected: string[]]} – the connected and disconnected relations ids
+ */
+export const diffRelations = (browserStateRelations = [], serverStateRelations = []) => {
+  const currentRelations = browserStateRelations.map((relation) => relation.id);
+  const oldRelations = serverStateRelations.map((relation) => relation.id);
+
+  const connected = currentRelations.filter((relation) => !oldRelations.includes(relation));
+  const disconnected = oldRelations.filter((relation) => !currentRelations.includes(relation));
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/RelationInputDataManager/utils/diffRelations.js:11]: Should we try and avoid looping twice here? I also find this easier to read ...

```suggestion
  const connected = browserStateRelations.filter((relation) => !oldRelations.includes(relation.id));
  const disconnected = serverStateRelations.filter((relation) => !currentRelations.includes(relation.id));
```

File: packages/core/admin/admin/src/content-manager/components/RelationInputDataManager/utils/normalizeRelations.js [hunk 10]
```diff
@@ -32,34 +32,13 @@ export const normalizeRelation = (relation, { shouldAddLink, mainFieldName, targ
...
-                  mainFieldName,
-                  targetModel,
-                })
-              )
-              .filter(Boolean)
-          )
-          ?.filter((page) => page.length > 0) ?? [],
-    },
-  };
+  return [...(relations ?? [])]?.map((relation) =>
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/RelationInputDataManager/utils/normalizeRelations.js:37]: Would it be possible to remove the conditional chaining here? `[...(relations ?? [])]` should always return an array, no?

File: packages/core/admin/admin/src/content-manager/hooks/useRelation/useRelation.js [hunk 11]
```diff
@@ -1,13 +1,20 @@
...
 
+export const useRelation = (cacheKey, { name, relation, search }) => {
+  const [searchParams, setSearchParams] = useState({});
+  const [currentPage, setCurrentPage] = useState(0);
+  /**
+   * This runs in `useInfiniteQuery` to actually fetch the data
+   */
   const fetchRelations = async ({ pageParam = 1 }) => {
     try {
+      setCurrentPage(pageParam);
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/hooks/useRelation/useRelation.js:16]: Should this be set BEFORE the call or after? In case an error occurs I think we do not want to set the page ...

File: packages/core/admin/admin/src/content-manager/hooks/useRelation/useRelation.js [hunk 12]
```diff
@@ -81,6 +90,37 @@ export const useRelation = (cacheKey, { relation, search }) => {
...
+  useEffect(() => {
+    if (status === 'success' && data && data.pages?.at(-1)?.results && onLoadRelationsCallback) {
+      // everytime we fetch, we normalize prior to adding to redux
+      const normalizedResults = normalizeRelations(data.pages?.at(-1)?.results, normalizeArguments);
+
+      // this is loadRelation from EditViewDataManagerProvider
+      onLoadRelationsCallback({
+        target: { name, value: normalizedResults },
+      });
+    }
```
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/hooks/useRelation/useRelation.js:119]: What benefit do you see to handle this as part of `useRelation()`? I'm wondering if the hook wouldn't be easier to use without `normalizeArguments` and `onLoadRelationsCallback`.

**Review by gu-stav — COMMENTED** [review]
Very well done! 👏🏼 One comment (about `initialData` and `modifiedData`) might be the only real issue, but there is good chance I misunderstood the test.

Overall it looks much cleaner and I'm very happy about the recursive helper you have introduced! ✨

File: packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/reducer.js (hunk 2 above)
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/reducer.js:173]: that's great, i think we need to get a "browserlist" to help us know when we can use these features 😄

File: packages/core/admin/admin/src/content-manager/hooks/useRelation/useRelation.js (hunk 12 above)
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/hooks/useRelation/useRelation.js:119]: I think it's better to have it here and run a callback because we never want the relations to not be normalised, a separation of concerns in the sense that `useRelations` should handle everything to do with them, it shouldn't know about the store _but_ it should fire a callback. Otherwise the effect is in the data manager.

I think I _wanted_ to stop returning the `relationsRes.data` because it's not used anyway, we only want the data to be in the store, hence the callback.

File: packages/core/admin/admin/src/content-manager/hooks/useRelation/useRelation.js (hunk 11 above)
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/hooks/useRelation/useRelation.js:16]: yep good point.

File: packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/tests/reducer.test.js (hunk 3 above)
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/tests/reducer.test.js:733]: No `modifedData` and `initialData` will still be different, this is also my problem with the naming, in reality `initialData` is the state of the server. Whilst `modifiedData` is it the state of the browser.

The flow would go like:
* init the edit view
* ID & MD are equal
* User makes an update
* MD is updated to reflect this
* User submits entry
* `cleanData` diffs ID & MD

File: packages/core/admin/admin/src/content-manager/components/RelationInput/RelationInput.js (hunk 5 above)
Comment (joshuaellis) [reply packages/core/admin/admin/src/content-manager/components/RelationInput/RelationInput.js:347]: Is this a ticket? This is something we can look at during `relations-reordering` and adjust accordingly, I can't imagine it would be crazy hard.

**Review by gu-stav — APPROVED** [review]
Let's go.
